### PR TITLE
added reset_idrac and known_good_state clean steps

### PR DESCRIPTION
**Pull request by kpradip001**
I have added reset_idrac and known_good_state clean steps and it's unit tests in ironic driver, please review it

### CODE REVIEW

File: ironic/drivers/modules/drac/raid.py [hunk 1]
```diff
@@ -769,6 +769,35 @@ def delete_configuration(self, task):
 
         return _commit_to_controllers(node, list(controllers))
 
+    @METRICS.timer('DracRAID.reset_idrac')
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/raid.py]: I think these cleaning steps should be in the DracManagement class instead of the RAID class.

File: ironic/drivers/modules/agent.py [hunk 2]
```diff
@@ -813,6 +813,32 @@ def _delete_configuration_final(task, command):
         task.node.raid_config = {}
         task.node.save()
 
+    @METRICS.timer('AgentRAID.reset_idrac')
```
Comment (cdearborn) [inline ironic/drivers/modules/agent.py]: We really shouldn't be making any changes to agent.py.  The cleaning steps that we're adding will be specific to our iDRAC driver.  You would only add the cleaning steps to agent.py if you were adding them for all drivers.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 3]
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
 
         self.assertEqual({'logical_disks': [expected_logical_disk]},
                          props)
+
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: If we remove the changes to agent.py, then these test cases can be removed.

File: ironic/tests/unit/drivers/modules/test_agent.py [hunk 4]
```diff
@@ -1662,7 +1662,11 @@ def test_get_clean_steps(self, get_steps_mock):
             {'step': 'create_configuration', 'interface': 'raid',
              'priority': 1},
             {'step': 'delete_configuration', 'interface': 'raid',
-             'priority': 2}]
+             'priority': 2},
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/test_agent.py]: If we remove the changes to agent.py, then these test cases can be removed.

**Comment by kpradip001** [comment]
Thanks for correcting me, I will do changes accordingly.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 5]
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
...
+                                  shared=False) as task:
+            return_value = task.driver.raid.reset_idrac(task)
+
+            mock_client.reset_idrac.assert_called_once_with(
+                force=True, wait=True)
+
+        self.assertIsNone(return_value)
+        self.node.refresh()
+
+    @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: It looks to me like this test is the same as the one above, but the one above checks the passed parameters.  If this is the case, can you remove this one?

File: ironic/tests/unit/drivers/modules/drac/test_raid.py [hunk 6]
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
...
+
+            mock_client.reset_idrac.assert_called_once_with(
+                force=True, wait=True)
+            mock_client.delete_jobs.assert_called_once_with(
+                job_ids=['JID_CLEARALL_FORCE'])
+
+        self.assertIsNone(return_value)
+        self.node.refresh()
+
+    @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_raid.py]: It looks to me like this test is the same as the one above, but the one above also checks the passed parameters.  If this is the case, can you remove this one?

File: ironic/drivers/modules/drac/raid.py (hunk 1 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/raid.py]: I have added cleaning steps in the DracManagement class in latest commit, can you please review it.

File: ironic/drivers/modules/agent.py (hunk 2 above)
Comment (kpradip001) [reply ironic/drivers/modules/agent.py]: I have corrected this one in latest commit.

File: ironic/tests/unit/drivers/modules/test_agent.py (hunk 4 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/test_agent.py]: I have removed this test case, as we don't makes any changes in agent.py.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 5 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: I have removed this test case, as we don't makes any changes in raid.py.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 6 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: I have removed this test case, as we don't makes any changes in raid.py.

**Comment by kpradip001** [comment]
I have addressed all review comments in latest commit, can you please review it?

File: ironic/drivers/modules/agent.py (hunk 2 above)
Comment (juliakreger) [reply ironic/drivers/modules/agent.py]: Driver management or bios interfaces please :)

Comment (kpradip001) [reply ironic/drivers/modules/agent.py]: I have implemented cleaning step in management interface, please check my latest commit.

File: ironic/conf/drac.py [hunk 7]
```diff
@@ -21,7 +21,13 @@
                min=1,
                help=_('Interval (in seconds) between periodic RAID job status '
                       'checks to determine whether the asynchronous RAID '
-                      'configuration was successfully finished or not.'))
+                      'configuration was successfully finished or not.')),
```
Comment (cdearborn) [inline ironic/conf/drac.py]: Hey Pradip, I don't think we want to be able to set the priority of these cleaning steps in ironic.conf for now.  Can you undo this change?

File: ironic/drivers/modules/agent.py (hunk 2 above)
Comment (cdearborn) [reply ironic/drivers/modules/agent.py]: Can you back out this change?  This is still showing up as a changed file in the PR.

File: ironic/drivers/modules/drac/management.py [hunk 8]
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
                   sensor type, which can be processed by Ceilometer.
         """
         raise NotImplementedError()
+
+    @METRICS.timer('DracManagement.reset_idrac')
+    @base.clean_step(priority=CONF.drac.clean_priority_reset_idrac)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: I think the priority for this step should be 0

File: ironic/drivers/modules/drac/management.py [hunk 9]
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
...
+        """Reset the iDRAC configuration.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
+        client.reset_idrac(force=True, wait=True)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:373]: I believe you need to tell ironic conductor to resume cleaning after the reset completes like this:
manager_utils.notify_conductor_resume_clean(task)

File: ironic/drivers/modules/drac/management.py [hunk 10]
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
...
+
+    @METRICS.timer('DracManagement.known_good_state')
+    @base.clean_step(priority=CONF.drac.clean_priority_known_good_state)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: I think we want this step to have a priority of 0

File: ironic/drivers/modules/drac/management.py [hunk 11]
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
...
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
+        client.reset_idrac(force=True, wait=True)
+        client.delete_jobs(job_ids=['JID_CLEARALL_FORCE'])
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:388]: I believe you need to tell ironic conductor to resume cleaning after the reset completes like this:
manager_utils.notify_conductor_resume_clean(task)

File: ironic/drivers/modules/drac/raid.py (hunk 1 above)
Comment (cdearborn) [reply ironic/drivers/modules/drac/raid.py]: Can you back out the changes to this file?  This file is still showing up as modified.

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 5 above)
Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Can you back out the changes to this file?  This file is still showing up as modified.

File: ironic/tests/unit/drivers/modules/test_agent.py (hunk 4 above)
Comment (cdearborn) [reply ironic/tests/unit/drivers/modules/test_agent.py]: Can you back out the changes to this file?  It's still showing up as modified.

**Review by cdearborn — CHANGES_REQUESTED** [review]
This looks pretty good Pradip!  There are a few files that we need to back out the changes for, and we need to add a line of code that tells ironic to resume cleaning when the cleaning steps have finished.

**Comment by kpradip001** [comment]
Thanks Chris, I will address mentioned review comments tomorrow.

File: ironic/drivers/modules/drac/raid.py (hunk 1 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/raid.py]: Done

File: ironic/drivers/modules/agent.py (hunk 2 above)
Comment (kpradip001) [reply ironic/drivers/modules/agent.py]: Done

File: ironic/tests/unit/drivers/modules/test_agent.py (hunk 4 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/test_agent.py]: Done

File: ironic/tests/unit/drivers/modules/drac/test_raid.py (hunk 5 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_raid.py]: Done

File: ironic/conf/drac.py (hunk 7 above)
Comment (kpradip001) [reply ironic/conf/drac.py]: Done

File: ironic/drivers/modules/drac/management.py (hunk 8 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py]: Done

File: ironic/drivers/modules/drac/management.py (hunk 10 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py]: Done

File: ironic/drivers/modules/drac/management.py (hunk 11 above)
Comment (kpradip001) [reply ironic/drivers/modules/drac/management.py:388]: I have added line :  agent_base_vendor._notify_conductor_reume_clean(task)  after delete jobs completes, but while executing cleaning steps, "Failed to tear down from cleaning" exception occurs at this line  -https://github.com/cdearborn/ironic/blob/master/ironic/conductor/manager.py#L1443 and node got stuck into CLEANING state.

Comment (cdearborn) [reply ironic/drivers/modules/drac/management.py:388]: I would try calling resume_cleaning the same way that the RAID creation cleaning step does here:
https://github.com/cdearborn/ironic/blob/master/ironic/drivers/modules/drac/raid.py#L903
I'm not sure if the call you are making is doing the same things or not.

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 12]
```diff
@@ -532,3 +532,31 @@ def test_get_sensors_data(self, mock_get_drac_client):
...
+        mock_get_drac_client.return_value = mock_client
+
+        with task_manager.acquire(self.context, self.node.uuid,
+                                  shared=False) as task:
+            return_value = task.driver.management.reset_idrac(task)
+            mock_client.reset_idrac.assert_called_once_with(
+                force=True, wait=True)
+
+            self.assertIsNone(return_value)
+            self.node.refresh()
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py]: The node.refresh() line isn't needed above.  It causes the node to be reloaded from the DB, which is only needed before doing asserts on properties of the node.

File: ironic/tests/unit/drivers/modules/drac/test_management.py [hunk 13]
```diff
@@ -532,3 +532,31 @@ def test_get_sensors_data(self, mock_get_drac_client):
...
+        with task_manager.acquire(self.context, self.node.uuid,
+                                  shared=False) as task:
+            return_value = task.driver.management.known_good_state(task)
+            mock_client.reset_idrac.assert_called_once_with(
+                force=True, wait=True)
+            mock_client.delete_jobs.assert_called_once_with(
+                job_ids=['JID_CLEARALL_FORCE'])
+
+            self.assertIsNone(return_value)
+            self.node.refresh()
```
Comment (cdearborn) [inline ironic/tests/unit/drivers/modules/drac/test_management.py]: The node.refresh() line isn't needed above.  It causes the node to be reloaded from the DB, which is only needed before doing asserts on properties of the node.

**Review by cdearborn — APPROVED** [review]
Looks great Pradip!  I did some poking and found this:
https://github.com/openstack/ironic/blob/master/ironic/drivers/base.py#L1452
So it does look like not returning anything is the right thing to do for a synchronous cleaning step.
I think you can safely remove the node.refresh() calls.  Try doing that and running the test cases, if you wouldn't mind.  If that all works, then please go ahead and submit the patch to ironic.
Thanks!

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 12 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py]: Done

File: ironic/tests/unit/drivers/modules/drac/test_management.py (hunk 13 above)
Comment (kpradip001) [reply ironic/tests/unit/drivers/modules/drac/test_management.py]: Done

**Comment by kpradip001** [comment]
> Looks great Pradip! I did some poking and found this:
> https://github.com/openstack/ironic/blob/master/ironic/drivers/base.py#L1452
> So it does look like not returning anything is the right thing to do for a synchronous cleaning step.
> I think you can safely remove the node.refresh() calls. Try doing that and running the test cases, if you wouldn't mind. If that all works, then please go ahead and submit the patch to ironic.
> Thanks!

Chris, I have addressed all review comments and created patch to ironic, this is link of patch https://review.opendev.org/#/c/658271/

**Comment by cdearborn** [comment]
Thanks Pradip!  I'm going ahead and closing this PR since it has now been submitted to ironic.
